src/UI/Actions.hs

```diff
@@ -0,0 +1,246 @@
+import qualified Brick.Widgets.List as L
+import Network.Mail.Mime (Address(..), renderSendMail, simpleMail')
+import Data.Semigroup ((<>))
+import Data.Vector (Vector)
+import Data.Text (unlines)
+import Data.Text.Lazy.IO (readFile)
+import Prelude hiding (readFile, unlines)
+import Control.Lens (set, over, view, (&), _Just)
+import Control.Lens.Fold ((^?!))
+import Control.Monad ((>=>))
+import Control.Monad.Except (runExceptT)
+import Control.Monad.IO.Class (liftIO)
+import Data.Text.Zipper (currentLine, gotoEOL)
+import Data.Text (Text)
+import Storage.Notmuch (getMessages, addTag, removeTag, setNotmuchMailTags)
+import Storage.ParsedMail (parseMail, getTo, getFrom, getSubject)
+import Types
+import Error
+
+backToIndex :: Action a
+backToIndex =
+    Action
+    { _aDescription = "back to the index"
+    , _aAction = continue . set asAppMode BrowseMail
+    }
+
+haltApp :: Action a
+haltApp =
+    Action
+    { _aDescription = "quit the application"
+    , _aAction = halt
+    }
+
+composeMail :: Action (L.List Name NotmuchMail)
+composeMail =
+    Action
+    { _aDescription = "compose a new mail"
+    , _aAction = continue . set asAppMode GatherHeaders
+    }
+
+focusSearch :: Action (L.List Name NotmuchMail)
+focusSearch =
+    Action
+    { _aDescription = "Manipulate the notmuch database query"
+    , _aAction = (continue
+                   . set asAppMode SearchMail
+                   . over (asMailIndex . miSearchEditor) (E.applyEdit gotoEOL))
+    }
+
+displayMail :: Action (L.List Name NotmuchMail)
+displayMail =
+    Action
+    { _aDescription = "display an e-mail"
+    , _aAction = \s ->
+                      do s' <-
+                             liftIO $
+                             updateStateWithParsedMail s >>=
+                             updateReadState removeTag
+                         continue s'
+    }
+
+setUnread :: Action (L.List Name NotmuchMail)
+setUnread =
+    Action
+    { _aDescription = "toggle unread"
+    , _aAction = (liftIO . updateReadState addTag >=> continue)
+    }
+
+applySearchTerms :: Action (E.Editor Text Name)
+applySearchTerms =
+    Action
+    { _aDescription = "apply search"
+    , _aAction = applySearch
+    }
+
+mailIndexUp :: Action (L.List Name NotmuchMail)
+mailIndexUp =
+    Action
+    { _aDescription = "mail index up one e-mail"
+    , _aAction = mailIndexEvent L.listMoveUp
+    }
+
+mailIndexDown :: Action (L.List Name NotmuchMail)
+mailIndexDown =
+    Action
+    { _aDescription = "mail index down one e-mail"
+    , _aAction = mailIndexEvent L.listMoveDown
+    }
+
+switchComposeEditor :: Action (L.List Name NotmuchMail)
+switchComposeEditor =
+    Action
+    { _aDescription = "switch to compose editor"
+    , _aAction = \s -> case view (asCompose . cTmpFile) s of
+                          Just _ -> continue $ set asAppMode ComposeEditor s
+                          Nothing -> continue s
+    }
+
+replyMail :: Action (L.List Name NotmuchMail)
```

Comment:
Does this action apply in the "view mail" mode too? Should it?  Again, not a blocker for
this PR.

Reply:
Yeah it doesn't, but good point.